### PR TITLE
blendid 5 - CSS and JS linting

### PR DESCRIPTION
**Pull request by benjtinsley**
since we frontend developers at viget use blendid on pretty much every project, having a good way to enforce our standards is desired. we spend a lot of time flagging pointless errors in PRs, and this is a step to mitigate that.

in blendid 5.0, we will ship with our standards set in `.stylelint` and `.eslint` files, making it easy for anyone to follow along with our standards, but also make it simple for people to configure it however they wish to match their own project requirements/preferences.

both stylelint and eslint will log out errors in the console when they happen. stylelint will auto fix certain errors (non-alphabetized CSS properties, incorrect indentation, etc). this also includes prettier, which, when adding the prettier plugin to your text editor, will auto fix certain errors (semicolons, incorrect indentation, etc).

## example of stylelint autofix (on save):
gif
### fixes displayed
- auto-alphabetizes properties
- adds line between nested declaration and parent properties
- corrects indentation

## example of prettier autofix (on save):
gif
### fixes displayed
- removes semicolons 
- corrects indentation 
- changes double quotes to single quotes

### CODE REVIEW

File: root/.stylelintrc [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+    "no-empty-source": true,
+    "no-extra-semicolons": true,
+    "rule-empty-line-before": [ "always", {
+      "except": ["first-nested"],
+      "ignore": ["after-comment"]
+    }],
+    "selector-pseudo-class-no-unknown": true,
+    "selector-pseudo-element-no-unknown": true,
+    "order/properties-alphabetical-order": true
+  }
```
Comment (benjtinsley) [inline root/.stylelintrc]: these rules are defined here: https://stylelint.io/user-guide/rules/
the plugin `stylelint-order` and its corresponding value of `"order/properties-alphabetical-order": true` allow us to auto alphabetize CSS properties

File: root/.eslintrc.js [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+    'no-console': 'off',
+    'no-debugger': isIntegration ? 'error' : 'off',
+    'no-unused-vars': [
+      1,
+      { args: 'none', ignoreRestSiblings: true, varsIgnorePattern: '^_' }
+    ],
+    'no-return-assign': 'error',
+    'prettier/prettier': isIntegration ? 'error' : 'warn'
+  }
+}
```
Comment (benjtinsley) [inline root/.eslintrc.js]: this is an abbreviated version of the one found in our react way: https://github.com/vigetlabs/our-react-way/blob/master/.eslintrc.js

File: package.json [hunk 3]
```diff
@@ -50,23 +55,30 @@
...
     "webpack-dev-middleware": "^1.12.0",
     "webpack-hot-middleware": "^2.18.2"
   },
   "devDependencies": {
     "chai": "^3.5.0",
     "mocha": "3.2.0"
+  },
+  "prettier": {
+    "semi": false,
+    "singleQuote": true
```
Comment (benjtinsley) [inline package.json]: prettier runs as a text editor plugin, so this will only work if users has this installed

File: .stylelintrc [hunk 4]
```diff
@@ -0,0 +1,4 @@
+{
+  "_developer-comment": "this file to stay empty to ensure CI builds pass. editable .stylelintrc found at root/.stylelintrc",
+  "rules": {}
+}
```
Comment (benjtinsley) [inline .stylelintrc]: kinda weird but the build command failed without this, hence CI builds failed without it. open for other suggestions of how to accomplish this

File: gulpfile.js/tasks/watch.js [hunk 5]
```diff
@@ -32,7 +27,7 @@ var watchTask = function() {
     if(taskConfig) {
       var srcPath = projectPath(PATH_CONFIG.src, taskPath.src)
       var globPattern = '**/*' + (taskConfig.extensions ? '.{' + taskConfig.extensions.join(',') + '}' : '')
-      watch(path.join(srcPath, globPattern), watchConfig, function() {
+      gulp.watch(path.join(srcPath, globPattern), watchConfig, function() {
```
Comment (benjtinsley) [inline gulpfile.js/tasks/watch.js]: i removed `gulp-watch` plugin since gulp has a default watch task.

File: gulpfile.js/tasks/watch.js [hunk 6]
```diff
@@ -12,10 +11,6 @@ var watchTask = function() {
         return PATH_CONFIG.icons
       case 'svgSprite':
         return PATH_CONFIG.icons
-      case 'html':
-        return PATH_CONFIG.html
-      case 'static':
-        return PATH_CONFIG.static
       default:
         return PATH_CONFIG[taskName]
```
Comment (benjtinsley) [inline gulpfile.js/tasks/watch.js]: the removed cases could be run with the default case. they were probably left there from a previous iteration that did not include the default case

File: root/.stylelintrc (hunk 1 above)
Comment (jeremyfrank) [reply root/.stylelintrc]: Here are the stylelint rules I had on Fashion Positive. Not sure if any of them would be of value to add here. https://github.com/C2CPII/fashion-positive/blob/master/.stylelintrc

Comment (benjtinsley) [reply root/.stylelintrc]: @jeremyfrank i dont have access to that repo. could you put it in a gist?

Comment (jeremyfrank) [reply root/.stylelintrc]: I'll just post here. It's not much.

```
{
  "rules": {
    "block-no-empty": true,
    "color-hex-case": "lower",
    "color-hex-length": "short",
    "comment-no-empty": true,
    "declaration-block-semicolon-newline-after": "always",
    "indentation": 2,
    "no-extra-semicolons": true,
    "number-leading-zero": "always",
    "number-no-trailing-zeros": true,
  }
}
```

**Pull request merged**